[P1][P2] Implement Components: Main, Argument Parser, API Keys Reader and Requester

deerhunter4 opened the pull request:
[Req][P2] Implement Requester Component

I created requester.py file. This component is responsible for
making requests to the API services and maintaining the response.

This branch should be used for program testing because it contains all the needed components.

deerhunter4 committed:
[Main][P1] Implement the Main component

I created main.py file that will be an entry point
to the program.
__pycache__ folder was also created.

deerhunter4 committed:
[Args][P1] Implement Argument Parser Component

I created arguments.py file. This component
is responsible for handling user input (parameters).

deerhunter4 committed:
[Keys][P1] Implement the API Keys File Reader Component

I created keys.py file. This component will be responsible for
handling user API keys stored in better_weather_keys.txt file.

deerhunter4 committed:
[Req][P2] Implement Requester Component

I created requester.py file. This component is responsible for
making requests to the API services and maintaining the response.

deerhunter4 committed:
Merge branch 'keys.p1.implement-keys' into req.p2.implement-requester

deerhunter4 committed:
I untracked __pycatche__ folder

KrzysiekJodlowski commented on requester.py:
```diff
@@ -0,0 +1,99 @@
+import requests
+
+# For now request works only for default '24h' forecast period
```
```suggestion
# Note: For now all requests work only for the default '24h' forecast period.
```

deerhunter4 replied:
done

KrzysiekJodlowski commented on requester.py:
```diff
@@ -0,0 +1,99 @@
+
+
+def request_weatherapi(keys_dict, args):
+    url_wheatherapi = 'https://api.weatherapi.com/v1/forecast.json'
+    API_KEY_wheatherapi = keys_dict['wheatherapi.com']
```
URLs and commonly-used variables should be constants.

deerhunter4 replied:
done for all URLs.

KrzysiekJodlowski replied:
I see, cool, the only thing is all the constants should be placed at the beginning of the file, before the methods. This way it can be reused by more than one method if needed. Please update that part and it'll be fine.

KrzysiekJodlowski replied:
Don't mind the recent comment - there are two kinds of variables you're using, URLs and keys. The first one I believe should be constants and grouped at the beginning of this file, Key values are passed to these methods and are regular parameters passed to the method. I see 3 options for the keys: 
- pass the whole dict and extract only the needed key
- pass the whole dict and pass only the needed key to the submethod
- pass only the needed key to the method already
You can pick whichever way you want.

deerhunter4 replied:
@KrzysiekJodlowski I extract all URLs from the functions and place them at the beginning of the script. As for the second issue, I think the code already matches the first option you proposed.

KrzysiekJodlowski commented on requester.py:
```diff
@@ -0,0 +1,99 @@
+# open-meteo.com API
+
+
+def request_openmeteo(latitude, longitude):
```
Why in this and the next function there's no error response handling?

deerhunter4 replied:
Yes, it is true. Honestly, I thought I could polish the code when the whole program worked as it should for basic parameters. I will get back to it in the future.

KrzysiekJodlowski replied:
I get your point but good practice would be to do it upfront, all this kind of polishing should be done at the implementation phase, especially since you already done it for one method. Let's discuss it at the upcoming meeting.

deerhunter4 replied:
@KrzysiekJodlowski I added error 403 handling to the request_meteostat() function. Request_openmeteo() function does not require an API key, so do not need error 403 handling.

KrzysiekJodlowski replied:
Ok, cool, as for the `request_openmeteo()` let it be if it works as you say. Still, I see the logic for handling the `403` is pretty much the same, can you please extract it to some common method? Also, it would be good to handle at least [the `503` error also](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503), so when the server is down. This can happen always e.g. because of some maintenance time. With having one method for error handling this should be a matter of adding just 3 other lines of code.

deerhunter4 replied:
@KrzysiekJodlowski Done. I created function **response_errors(**api_response, name**)** that checks 403 and 503 errors. I used it in all three API request functions.

KrzysiekJodlowski replied:
@deerhunter4 sounds good but I cannot see any code changes?

deerhunter4 replied:
@KrzysiekJodlowski Now you should see

KrzysiekJodlowski commented on requester.py:
```diff
@@ -0,0 +1,99 @@
+def request_meteostat(keys_dict, latitude, longitude, current_date):
+    url_meteostat = 'https://meteostat.p.rapidapi.com/point/hourly'
+    API_KEY_meteostat = keys_dict['meteostat.net']
+    # current_date = hours_wheatherapi[1].split(' ')[0]
+
+    params_meteostat = {
+        "lat": latitude,
+        "lon": longitude,
+        "start": current_date,  # current day obtained from first api request
```
Why the `current_date` does not work in this case? In general, it makes sense to pass it over to all those functions because the user chooses it. Even if it's hardcoded to 24h as for now, please hardcode it in the Main component and pass it to those. This way you'll be ready for the future.

deerhunter4 replied:
I created current_date variable, but I am using it only for the last API request. I checked other API services but I did not find a way to add it to the other's requests easily. I will deal with it later.

KrzysiekJodlowski replied:
Again, often things like that left for later are not done or easily forgotten. Here it's a matter of passing one arg to all the functions, it should not be complicated. Let's talk about it at the upcoming meeting.

deerhunter4 replied:
@KrzysiekJodlowski I could pass current_date as an argument to the other functions responsible for API requests, but it would not be used, at least in 24h period forecast. I have not found such a parameter in other API services so far. I will deal with it later when it will be needed.

KrzysiekJodlowski replied:
@deerhunter4 ok it will not be used in the default `24h` period case, but what if the user chooses to use another period time? For E.g. in the `request_openmeteo` method, I see you use the `forecast_days` param in the request, so this could and should be doable. And even if you don't want to use that param now (which is in contrary to the requirements we have, and that requires yet another ticket for handling that issue later), please at least pass it for now with a comment saying this part will be dealt with later (and let me know explicitly so that I will create a task for it). Again:

- let's talk about it at the upcoming meeting if you're not sure
- before the meeting please at least try to handle that, and if not successful, pass at least a placeholder param for now with a comment
- we can also try to figure that out together if you want and by that occasion practice some [pair programming](https://en.wikipedia.org/wiki/Pair_programming)
- all that because, as I already said "things like that left for later are not done or easily forgotten";)

deerhunter4 replied:
@KrzysiekJodlowski I think that pair programming was a success. Everything looks ready for a merge.

KrzysiekJodlowski reviewed (changes requested):
Left some comments, also, all three functions seem to be a bit longish, is there an easy way to refactor it somehow and e.g. extract some common operation into helper functions? You can e.g. use the comments you left to decide on that.

deerhunter4 committed:
Change in main.py and README.md files
All 'wheather' typo was corrected to 'weather' in all the files.
In README.md example of the better_weather_keys.txt file was
formatted as code block.
In main.py file parameters and main sections were updated.

deerhunter4 committed:
Simplyfying main.py by getting longitude and latitude from function.
I slightly modify request_weatherapi() function in requester.py
to get as return also longitude and latitude.
Current_date variable is now obtained independently
from the function.

deerhunter4 committed:
main.py and keys.py files were modified.
read_keys() function, which is used by check_keys() function
was created in keys.py.
In main.py was created KEYS_PATH constant.

deerhunter4 commented:
@KrzysiekJodlowski 
I could not find any code that might be refactored easily. I think that in the future when I will develop API requests for other time periods code refactoring will be easier. 
And as I wrote before all comments from other Pull Requests were addressed but if they involved code changes they all were done in this Pull Request.
I hope that you will be satisfied by my responses and merge the branches quickly.

deerhunter4 committed:
Small changes in requester.py
Some constan variables were added (URLs)

KrzysiekJodlowski commented:
@deerhunter4 thanks for addressing all (or, almost all) comments. Still, I think some of the leftovers should be taken care of sooner rather than later. Also, since you're addressing all those in this one PR you should now close all the remaining PRs and rename that one, to reflect what tickets are covered here.

deerhunter4 committed:
[Req][P2] Constant URL variables moved to the begining of requester.py

deerhunter4 committed:
[Req][P2] Error 403 handling was added to the request_meteostat() function

deerhunter4 committed:
[P1][P2] Unification of error messages.
Error messages were unifcated in components: keys.py,
arguments.py and requester.py.
Additionally new error handling sections were add
to the requester.py component.

deerhunter4 commented:
@KrzysiekJodlowski I addressed all the comments. I hope that you are satisfied with my answers. If yes, please merge the pull request.

KrzysiekJodlowski commented on requester.py:
```diff
@@ -0,0 +1,104 @@
+import requests
+
+URL_WEATHERAPI = 'https://api.weatherapi.com/v1/forecast.json'
+URL_OPENMETEO = "https://api.open-meteo.com/v1/forecast"
+URL_METEOSTAT = 'https://meteostat.p.rapidapi.com/point/hourly'
```
Please be consistent with the quote usage, use either `'` or `"`.

deerhunter4 replied:
Done

KrzysiekJodlowski replied:
Cool, but I cannot see any code changes?

deerhunter4 replied:
Now you should see.

KrzysiekJodlowski reviewed (changes requested):
@deerhunter4 thank you for doing all the updates, good job so far. Still, the error and `current_date` param handling for me lacks some polishing. Apart from that, three remarks from my side:

- remember about re-requesting the review if you want  the reviewer to take a look at the PR again
- regarding:
>If yes, please merge the pull request

Handling the code side, including PRs and its merging is the PR requester responsibility.
- I see you have moved all the related tickets to "Testing" already, you should not do it if the code is not merged yet. Please bring them back to the relevant column.

Overall, I suggest scheduling some longer sync-up meeting next week to deal with the remaining issues, there are not many things to be done, but doing them together should speed up things. There is also one more thing I'm missing here, I cannot see any unit tests, so let's look at this code together and see if we can do anything about it.

deerhunter4 committed:
[Req][P2] Quotation marks have been unified.

deerhunter4 committed:
[Req][P2] Function response_errors() was created in requester.py
New function check if lack of proper request response
is due to 403 or 503 error.
All three API request functions use the new function.

KrzysiekJodlowski commented:
@deerhunter4 please also remove the API_request_test.py file within this PR. Please also add [the .gitignore file](https://www.atlassian.com/pl/git/tutorials/saving-changes/gitignore) to ignore the pycache files and any other files that are not needed anymore, like API_request_test.py.

deerhunter4 committed:
[Args][P1] Data parameter added to arguments.py
check_date() function was also added to the file,
it is checking if date format is correct

deerhunter4 committed:
.gitignore file was created
Start ignoring: API_request_test.py and pycache files

deerhunter4 committed:
[Main][P1]"__name__" code part moved to the beginning of a file

KrzysiekJodlowski commented:
@deerhunter4 let's talk about this PR in private, I still see significant room for improvement and I don't believe we should merge it as it is now.

deerhunter4 committed:
[P1] .gitignore upadated

deerhunter4 committed:
[P1][README] Add Preconditions and Usage sections

deerhunter4 committed:
[P1] Component names update

deerhunter4 committed:
[P1][Args] Refactor argument_parser component

deerhunter4 committed:
[P1] Update components names and refactor argument_parser

deerhunter4 committed:
[P1][Keys] Refactor key_reader component

deerhunter4 committed:
[P2][Req] Refactor api_requester component

Code has been separated from three request functions
to create three new parser functions.
Some names of inner variables were changed.
Names were unified in all the functions.

KrzysiekJodlowski commented:
The code was reviewed and refactored together with @deerhunter4 to make sure it's in an acceptable state as well as to have a knowledge-sharing session and practice pair programming.

KrzysiekJodlowski reviewed (approved):
Good work @deerhunter4!👏🎉

Please remember to do a proper cleanup on our task board after the merge is done.
